Charging Algortihm.c

```diff
@@ -1,21 +1,73 @@
 #include <math.h>
-#define BMS_CAN_ID 0x1806E5F4
-#define CCS_CAN_ID 0x18FF50E5
+#include <stdint.h>
+#include <stdbool.h>
+
+#define BMS_CAN_ID 0x1806E5F4 // ID of message we send to Charger with target Voltage, Current and charging state
+#define CCS_CAN_ID 0x18FF50E5 // ID of message we should recive from Charger with it's current Temperature, Charging Current, Charging Voltage and SoC 
+
 #define SOC_THRESHOLD 90
 #define CAN_MSG_DLC 8
 #define CHARGING_STATE_OFF 1
 #define CHARGING_STATE_ON 0
 
-#define MAX_VOLTAGE_HIGH_BYTE 0
-#define MAX_VOLTAGE_LOW_BYTE 1
-#define MAX_CURRENT_HIGH_BYTE 2
-#define MAX_CURRENT_LOW_BYTE 3
-#define CHARGING_STATE 4
+/*Data indexes for message BMS sends to Charger*/
+#define MAX_VOLTAGE_HIGH_BYTE_INDEX 0
+#define MAX_VOLTAGE_LOW_BYTE_INDEX 1
+#define MAX_CURRENT_HIGH_BYTE_INDEX 2
+#define MAX_CURRENT_LOW_BYTE_INDEX 3
+#define CHARGING_STATE_INDEX 4
+
+/*Data indexes for message BMS receives from Charger*/
+#define DATA_VOLTAGE_HIGH_BYTE_INDEX 0
+#define DATA_VOLTAGE_LOW_BYTE_INDEX 1
+#define DATA_CURRENT_HIGH_BYTE_INDEX 2
+#define DATA_CURRENT_LOW_BYTE_INDEX 3
+#define DATA_STATUS_FLAGS_INDEX 4 
+
+/*Indexes of flags received form Charger*/
+#define HARDWARE_FAILURE_FLAG_INDEX 0
+#define TEMPERATURE_OF_CHARGER_FLAG_INDEX 1
+#define INPUT_VOLTAGE_FLAG_INDEX 2
+#define STARTING_STATE_FLAG_INDEX 3
+#define COMMUNICATION_STATE_FLAG_INDEX 4
+
+/*Definition of flags*/
+/*Hardware Faliure flags*/
+#define NORMAL 0
+#define HARDWARE_FALIURE 1
+/*Temperature of Charger flags*/
+#define NORMAL_TEMPERATURE 0
+#define OVER_TEMPERATURE_PROTECTION 1
+/*Input Voltage flags*/
+#define NORMAL_INPUT_VOLTAGE 0
+#define WRONG_INPUT_VOLTAGE 1
+/*Starting State flags*/
+#define BATTERY_DETECTED 0
+#define BATTERY_NOT_DETECTED 1
+/*Communication State flags*/
+#define NORMAL_COMMUNICATION 0
+#define COMMUNACTION_TIMED_OUT 1
+
+bool flagsArray[8]; //flag array is the size of the one byte and consits only of bits
```

Comment:
u komentaru si lipo objasnia i onda stavia da se flagsArray sastoji od 8 bajtova

Reply:
mozes stavit varijablu uint8_t flags jer imas jedan bajt, doli si dobro to zamislia sa onim provjerama ali nemos bas ovo sa indexima risit. Kad ides provjerit jeli se neki flag u ovom "flags-u" radis &(AND) operaciju na nacin:

if ((flags & STARTING_STATE_FLAG_INDEX) == STARTING_STATE_FLAG_INDEX)
{
STARTING_STATE_FLAG_INDEX je postavljen;
}
else{
tog flaga nema
}

sta se odi dogodi, flags ti je npr. 1011 1111, a STARTING_STATE_FLAG_INDEX je 0000 0100. Rezutlat AND operacije ce bit 0000 0100 sta odgovara flagu kojeg provjeravamo. Da ga nema dobia bi 0000 0000.

flagovi ti ne funkiconiraju ko indexu u arrayu jer oznacavaju bitove a ne bajtove.  Ne postoji nacin da ti pristupis bitu preko indexa ko za array, nego taj flag(bit) prikazes brojen, evo ti sad primjer:

0 - 0000 0000 nula flag
1 - 0000 0001 prvi bit
2 - 0000 0010 drugi bit
4 - 0000 0100 treci bit
8 - 0000 1000 cetvrti bit
itd...

tako da ti je gore pravilnije napisat da su tekonstante samo flag a ne index, STARTING_STATE_FLAG ti je onda vrijednost 4 a ne 3 jer 3 ti je (0000 0011), flag nesmi imat dvi jedinice.

Flagovi se koriste u embeddedu da bi ustedia na memoriji, u nasen konkretnon slucaju mi smo za 5 razlicith stanja uredaja iskoristili 1 bajt, a mogli smo bit rasipni i za svakoga koristit 5 bajtova. Npr mi sad imamo 32bitni miktronoler sta znaci da imamo 2^32 adresa pa se ne moramo bojat, a da imas 8bitni mikrokontoler vec bi se tribalo pazit na memoriju jer bi tad ima samo 256 dostupnih adresa u RAM-u.